chips/stm32f3xx/src/usart.rs

```diff
@@ -0,0 +1,464 @@
+        /// Overrun error clear flag
+        ORECF OFFSET(3) NUMBITS(1) [],
+        /// Noise detected clear flag
+        NCF OFFSET(2) NUMBITS(1) [],
+        /// Framing error clear flag
+        FECF OFFSET(1) NUMBITS(1) [],
+        /// Parity error clear flag
+        PECF OFFSET(0) NUMBITS(1) []
+    ]
+];
+
+const USART1_BASE: StaticRef<UsartRegisters> =
+    unsafe { StaticRef::new(0x40013800 as *const UsartRegisters) };
+const USART2_BASE: StaticRef<UsartRegisters> =
+    unsafe { StaticRef::new(0x40004400 as *const UsartRegisters) };
+const USART3_BASE: StaticRef<UsartRegisters> =
+    unsafe { StaticRef::new(0x40004800 as *const UsartRegisters) };
+
+pub struct Usart<'a> {
+    registers: StaticRef<UsartRegisters>,
+    clock: UsartClock,
+
+    tx_client: OptionalCell<&'a dyn hil::uart::TransmitClient>,
+    rx_client: OptionalCell<&'a dyn hil::uart::ReceiveClient>,
+    // tx_len: Cell<usize>,
+    // rx_len: Cell<usize>,
+}
+
+pub static mut USART1: Usart = Usart::new(
+    USART1_BASE,
+    UsartClock(rcc::PeripheralClock::APB2(rcc::PCLK2::USART1)),
+);
+
+pub static mut USART2: Usart = Usart::new(
+    USART2_BASE,
+    UsartClock(rcc::PeripheralClock::APB1(rcc::PCLK1::USART2)),
+);
+
+pub static mut USART3: Usart = Usart::new(
+    USART3_BASE,
+    UsartClock(rcc::PeripheralClock::APB1(rcc::PCLK1::USART3)),
+);
+
+impl Usart<'a> {
+    const fn new(base_addr: StaticRef<UsartRegisters>, clock: UsartClock) -> Usart<'a> {
+        Usart {
+            registers: base_addr,
+            clock: clock,
+
+            tx_client: OptionalCell::empty(),
+            rx_client: OptionalCell::empty(),
+            // tx_len: Cell::new(0),
+            // rx_len: Cell::new(0),
+        }
+    }
+
+    pub fn is_enabled_clock(&self) -> bool {
+        self.clock.is_enabled()
+    }
+
+    pub fn enable_clock(&self) {
+        self.clock.enable();
+    }
+
+    pub fn disable_clock(&self) {
+        self.clock.disable();
+    }
+
+    // for use by panic in io.rs
+    pub fn send_byte(&self, byte: u8) {
+        // loop till TXE (Transmit data register empty) becomes 1
+        while !self.registers.isr.is_set(ISR::TXE) {}
+        self.registers.tdr.set(byte.into());
+    }
+
+    fn enable_transmit_complete_interrupt(&self) {
+        self.registers.cr1.modify(CR1::TCIE::SET);
+    }
+
+    fn disable_transmit_complete_interrupt(&self) {
+        self.registers.cr1.modify(CR1::TCIE::CLEAR);
+    }
+
+    fn clear_transmit_complete(&self) {
+        self.registers.isr.modify(ISR::TC::CLEAR);
+    }
+
+    pub fn handle_interrupt(&self) {
+        self.clear_transmit_complete();
+        self.disable_transmit_complete_interrupt();
+
+        debug!("transmit");
+
+        self.enable_transmit_complete_interrupt();
+    }
+}
+
+impl hil::uart::Transmit<'a> for Usart<'a> {
+    fn set_transmit_client(&self, client: &'a dyn hil::uart::TransmitClient) {
+        self.tx_client.set(client);
+    }
+
+    // TODO
```

Comment:
Need implementation

Reply:
I stil need some time to figure out how DMA works on F3, will do asap.

Reply:
Done, for now without DMA, using interrupts.